### ISSUE TITLE
Inconsistent Tpetra options with TPL_ENABLE_CUDA=ON

### ISSUE THREAD

**Issue by tmdelellis** [trilinos/Trilinos#1939]
In `tpetra/CMakeLists.txt` the following checks (implemented in #1682) are too restrictive:

```
  29 IF (TPL_ENABLE_CUDA)
  30   # As of 10 Aug 2017, Tpetra requires CUDA >= 7.5.
  31   IF (DEFINED CUDA_VERSION)
  32     # Don't check a variable's value unless we know it is defined.
  33     IF (CUDA_VERSION VERSION_LESS "7.5")
  34        MESSAGE(FATAL_ERROR "If building with CUDA, Tpetra requires at least CUDA 7.5, and preferably      CUDA >= 8.0.  Your CUDA_VERSION is ${CUDA_VERSION}.  For details, please refer to Trilinos issue #1     278: https://github.com/trilinos/Trilinos/issues/1278")
  35     ENDIF ()
  36   ENDIF ()
  37
  38   IF (NOT DEFINED Kokkos_ENABLE_Cuda_Lambda OR NOT Kokkos_ENABLE_Cuda_Lambda)
  39     MESSAGE (FATAL_ERROR "If building with CUDA, Tpetra requires that you set Kokkos_ENABLE_Cuda_Lam     bda:BOOL=ON, if it is not already ON by default.  For details, please refer to Trilinos issue #1682      (https://github.com/trilinos/Trilinos/issues/1682).  You should also make sure, at least with CUDA_V     ERSION 7.5 or 8.0, to add \"--expt-extended-lambda\" to CMAKE_CXX_FLAGS.")
  40   ENDIF ()
  41 ENDIF ()
```

This effectively precludes settings `TPL_ENABLE_CUDA` with `Kokkos_ENABLE_Cuda=OFF`.  This, in my mind, can be a common use-case.  Several applications may be using CUDA for their own concurrency but using Trilinos for host-side solvers. In this case, one would only want Trilinos to run on the host, hence turning Kokkos_Cuda off (to avoid potential collisions on device).  

I believe the solution to the above is to simply check for CUDA Kokkos:
```
IF (Kokkos_ENABLE Cuda AND (NOT DEFINED Kokkos_ENABLE_Cuda_Lambda OR NOT Kokkos_ENABLE_Cuda_Lambda))
```

**Comment by aprokop** [comment]
@trilinos/tpetra

**Comment by mhoemmen** [comment]
@tmdelellis I'm calling this an enhancement, not a bug, since it would be an unusual use case to build Trilinos with CUDA enabled, yet disable CUDA in Kokkos.  It takes some effort to do that.  Nevertheless, I'll be happy to accept a pull request if you're willing to test it.  Thanks!

**Comment by tmdelellis** [comment]
@mhoemmen This is actually our standard build case.  We've been using this configuration for some time.  Only recently has this issue shown up.  To get around it we are currently setting `TPL_ENABLE_CUDA=ON` `Kokkos_ENABLE_Cuda=OFF` and `Kokkos_ENABLE_Cuda_Lambda=ON` .  It is the last statement, added to avoid a configuration error in Tpetra, that makes no sense.  If Kokkos Cuda is not on, you should not have to defined Kokkos CUDA Lambda just to get past a check in tpetra.

**Comment by mhoemmen** [comment]
@tmdelellis Please post a pull request; we'll be happy to review it.  Thanks!

**Comment by kddevin** [comment]
@crtrott How hard would it be for us to check the fix proposed above by @tmdelellis ?  Could we easily work it into one of your existing test builds?

**Comment by mhoemmen** [comment]
@kddevin The issue is that it requires a build that has CUDA enabled in Kokkos, but disables CUDA _instantiations_ (yet builds with CUDA) in Trilinos.  CUDA builds are much slower than non-CUDA builds, so this Dashboard build would waste resources that would be better spent on CUDA-enabled build combinations.  This is why I would be happy to take a pull request or at least hear that the proposed fix was tested for the desired use case.  "I believe it works" != "I tested it and it works."

Compare to Tpetra's support for `GO=unsigned long`, whose status is "some app wanted this and agreed to take on the burden of testing it."  I don't think anyone else tests that case by default.

**Comment by kddevin** [comment]
Perhaps it could be exercised once in @crtrott 's tests more easily than having another Tpetra team member create a brand new build to test a pull request.  I did not intend that it should be a new nightly test on the Dashboard.

**Comment by mhoemmen** [comment]
@kddevin that's fine.

**Comment by mhoemmen** [comment]
@tmdelellis Pull request #1961 purports to fix this issue.  Could you please take a look and see if it would actually fix it for your use case?  Thanks!

**Comment by mhoemmen** [comment]
I found some minor bugs in #1961.  If this works, then I'll just push to develop and delete that PR.

**Comment by mhoemmen** [comment]
I'm at a conference here in town this week and had a little time to work on this.  I fixed the issue (in a more general way), and tested Tpetra configure, build, and test.  I tested two configurations, both with CUDA enabled (`TPL_ENABLE_CUDA:BOOL=ON`).  The first is the usual one, with `Kokkos_ENABLE_Cuda` and `Tpetra_INST_CUDA` both ON.  The second is the use case that @tmdelellis wishes to work, namely with both `Kokkos_ENABLE_Cuda` and `Tpetra_INST_CUDA` both OFF.  Both configurations build and pass tests in TpetraCore.

I am running the check-in test now.
